src/gpt-auto-generator/gpt-auto-generator.c

```diff
@@ -526,43 +567,46 @@ static int add_partition_esp(DissectedPartition *p, bool has_xbootldr) {
                 return 0;
         }
 
-        /* If /efi exists we'll use that. Otherwise we'll use /boot, as that's usually the better choice, but
-         * only if there's no explicit XBOOTLDR partition around. */
-        if (access("/efi", F_OK) < 0) {
-                if (errno != ENOENT)
-                        return log_error_errno(errno, "Failed to determine whether /efi exists: %m");
-
-                /* Use /boot as fallback, but only if there's no XBOOTLDR partition and /boot exists */
-                if (!has_xbootldr) {
-                        if (access("/boot", F_OK) < 0) {
-                                if (errno != ENOENT)
-                                        return log_error_errno(errno, "Failed to determine whether /boot exists: %m");
-                        } else {
+        /* If /boot/ is present, unused, and empty, we'll take that.
+         * Otherwise, if /efi/ is unused and empty (or missing), we'll take that.
+         * Otherwise, we do nothing.
+         */
+        if (!has_xbootldr && slash_boot_exists()) {
+                r = slash_boot_in_fstab();
+                if (r < 0)
+                        return r;
+                if (r == 0) {
+                        r = path_is_busy("/boot");
+                        if (r < 0)
+                                return r;
+                        if (r == 0) {
                                 esp_path = "/boot";
                                 id = "boot";
                         }
                 }
         }
-        if (!esp_path)
+
+        if (!esp_path) {
+                r = RET_NERRNO(access("/efi", F_OK));
+                if (r < 0 && r != -ENOENT)
+                        return log_error_errno(r, "Failed to determine whether /efi/ exists: %m");
```

Comment:
you are not actually using the result of this test?

Reply:
It checks that the dir is accessible. But I guess we don't need to do that. I'll drop it.